feat(NgComponentOutlet): add NgComponentOutlet directive

mhevery opened the pull request:
feat(NgComponentOutlet): add NgComponentOutlet directive

Add NgComponentOutlet directive that can be used to dynamically create 
host views from a supplied component.

Closes #11168
Takes over PR #11235

shlomiassaf committed:
feat(NgComponentOutlet): add NgComponentOutlet directive

Add NgComponentOutlet directive that can be used to dynamically create
host views from a supplied component.

Closes #11168
Takes over PR #11235

shlomiassaf commented:
@mhevery I have pushed the updated code to my branch.

Tests are passing.

It includes a rebase +  all the requested changed except the namings.

Commits should be squashed :)

There are some lines of codes commented out in the tests since **Output** is not supported in * syntax.

I think you should use it as a baseline

https://github.com/shlomiassaf/angular/tree/ng-component-outlet

mhevery committed:
feat(NgTemplateOutlet): Make NgTemplateOutlet compatible with * syntax

BREAKING CHANGE:

- Deprecate `ngOutletContext`. Use `ngTemplateOutletContext` instead

mhevery committed:
fix: correctly show error when karma fails to load

mhevery commented:
@shlomiassaf thanks, i am also making sure that `ngTemplateOutlet` and `ngComponentOutlet` are consistent and adding documentation. At this point I think the two have diverged too much to be able to reuse your branch. Sorry.

mhevery commented:
@vicb ready for review.

MaximCrabbe commented:
So how long does it take on average for a PR like this to be reviewed? (Just would like to know, not 'pushing'). This NgComponentOutlet seems usefull!

shlomiassaf commented:
Holiday season I guess (:

Sent from my iPhone

> On Dec 30, 2016, at 3:48 PM, Maxim Crabbé <notifications@github.com> wrote:
> 
> So how long does it take on average for a PR like this to be reviewed? (Just would like to know, not 'pushing'). This NgComponentOutlet seems usefull!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

vicb commented on modules/@angular/common/src/directives/ng_template_outlet.ts:
```diff
@@ -15,42 +15,47 @@ import {Directive, EmbeddedViewRef, Input, OnChanges, SimpleChanges, TemplateRef
-  private _templateRef: TemplateRef<any>;
 
-  constructor(private _viewContainerRef: ViewContainerRef) {}
+  @Input() public ngTemplateOutletContext: Object;
```
What about `@Input('ngTemplateOutletContext') public context: Object;`

pkozlowski-opensource replied:
IMO it is better as-is

vicb commented on modules/@angular/common/src/directives/ng_template_outlet.ts:
```diff
@@ -15,42 +15,47 @@ import {Directive, EmbeddedViewRef, Input, OnChanges, SimpleChanges, TemplateRef
 
-  @Input()
-  set ngOutletContext(context: Object) { this._context = context; }
+  @Input() public ngTemplateOutlet: TemplateRef<any>;
```
What about `@Input('ngTemplateOutlet') public template: TemplateRef<any>;`

pkozlowski-opensource replied:
IMO it is better as-is

vicb commented:
LGTM, please rebase and consider the minor comments.

mhevery commented:
Rebased

pkozlowski-opensource commented on modules/@angular/common/src/directives/ng_component_outlet.ts:
```diff
@@ -0,0 +1,82 @@
+      private _viewContainerRef: ViewContainerRef) {}
+
+  ngOnChanges() {
+    this._viewContainerRef.clear();
```
Could we change it so only views we've created are destroyed here? In theory we could have multiple directives sitting on the same `<template>` element, they could be creating their own views and this logic would mess up views created by other directives, no?

This is what is done in the template outlet: https://github.com/angular/angular/blob/35f9a1c2cb743d050e000cb8865a70240ce3ed6d/modules/%40angular/common/src/directives/ng_template_outlet.ts#L48-L50

mhevery replied:
good point. Fixed

lgleim commented on modules/@angular/examples/common/ngComponentOutlet/ts/module.ts:
```diff
@@ -0,0 +1,83 @@
+// #docregion CompleteExample
+@Injectable()
+class Greeter {
+  suffix = '!'
```
[32, 15]: Missing semicolon.

lgleim commented on modules/@angular/common/src/directives/ng_component_outlet.ts:
```diff
@@ -0,0 +1,86 @@
+ * Instantiates a single {@link Component} type and inserts its Host View into current View.
+ * `NgComponentOutlet` provides a declarative approach for dynamic component creation.
+ *
+ * `NgComponentOutlet` requires a component type, if a falsy value is set the view will clear and
```
Should probably be `if a false value` instead of `if a falsy value`.

matsko commented:
@mhevery please fix CI issues.

matsko commented:
Landed as https://github.com/angular/angular/commit/8578682dcfccc2ccccb68df37c3ebc738f98bf0e

DzmitryShylovich commented:
beta 4
